### DISCUSSION TITLE
C++ Coroutines / Async + TLS Session Ticket Callbacks

### DISCUSSION THREAD

**Discussion by victorstewart** [openssl/openssl#24634] [Q&A]
i'm exploring design strategies to implement datacenter wide single use TLS session resumption tickets (so that if a connection breaks and client get rerouted to a new application server process the connection can be resumed, chiefly to support 0-RTT data).

ticket generation is fine the callbacks can fall-through to a coroutine that asynchronously writes the ticket to the database. but on ticket validation, evaluating whether a ticket has been used on any other server process and thus revoked, this would require an asynchronous database read/delete-if inside the callback.

in theory this should be possible since the state for each connection is isolated/local to a `SSL *` object, but i figured i'd ask whether this would work before I go down the path of implementing it.

and if this design is a no-go, maybe there's some other API that can be exposed to accomplish this asynchronously?

**Comment by nhorman** [comment]
Marking as inactive, to be closed when 3.4 dev cycle ends, barring further input

**Comment by victorstewart** [comment]
would still love feedback on this

**Comment by nhorman** [comment]
@victorstewart ok, sorry, didn't think a two year old issue would still have someone paying attention to it.

to restate your request, If I read you right, you want to:
1) use the SSL_set_session_ticket_ext_cb to catch ticket extensions, and add their information to a shared database between disparate serving processes on your server
2) On connection to a given serving process, said process looks up the ticket in the database
3) If found, use the ticket info to rebuild the session from the old process and resume

I'm speaking a bit naively here, but, In principle, I don't see anything wrong with doing so, but I don't immediately see the api calls needed to rebuiild the session based on ticket info alone.  We have the ability to duplicate sessions that are resumeable, but they are internal and require access to the entire old SSL_SESSION data structure to duplicate the session for resumption, so this seems incumbent on having access to all the old data to accomplish that.

Its an interesting idea though.  Its feels like you would need some serialization method that would let you export the entire SSL_SESSION data and store it in the data base to rebulid the session cache on your distributed systems

Note, I'm migrating this to a discussion, since its not a specific bug.

**Reply by victorstewart** [reply]
hey sorry just wanted to say i've been dealing with a family emergency the last few weeks, but will circle back once I'm back at work.
